pkg/kubectl/cmd/wait/wait.go

```diff
@@ -0,0 +1,333 @@
+		return nil, err
+	}
+	builder := flags.ResourceBuilderFlags.ToBuilder(flags.RESTClientGetter, args)
+	clientConfig, err := flags.RESTClientGetter.ToRESTConfig()
+	if err != nil {
+		return nil, err
+	}
+	dynamicClient, err := dynamic.NewForConfig(clientConfig)
+	if err != nil {
+		return nil, err
+	}
+	conditionFn, err := conditionFuncFor(flags.ForCondition)
+	if err != nil {
+		return nil, err
+	}
+
+	o := &WaitOptions{
+		ResourceFinder: builder,
+		DynamicClient:  dynamicClient,
+		Timeout:        flags.Timeout,
+
+		Printer:     printer,
+		ConditionFn: conditionFn,
+		IOStreams:   flags.IOStreams,
+	}
+
+	return o, nil
+}
+
+func conditionFuncFor(condition string) (ConditionFunc, error) {
+	if strings.ToLower(condition) == "delete" {
+		return IsDeleted, nil
+	}
+	if strings.HasPrefix(condition, "condition=") {
+		conditionName := condition[len("condition="):]
+		return ConditionalWait{
+			conditionName: conditionName,
+			// TODO allow specifying a false
+			conditionStatus: "true",
+		}.IsConditionMet, nil
+	}
+
+	return nil, fmt.Errorf("unrecognized condition: %q", condition)
+}
+
+// WaitOptions is a set of options that allows you to wait.  This is the object reflects the runtime needs of a wait
+// command, making the logic itself easy to unit test with our existing mocks.
+type WaitOptions struct {
+	ResourceFinder ResourceFinder
+	DynamicClient  dynamic.Interface
+	Timeout        time.Duration
+
+	Printer     printers.ResourcePrinter
+	ConditionFn ConditionFunc
+	genericclioptions.IOStreams
+}
+
+// ConditionFunc is the interface for providing condition checks
+type ConditionFunc func(info *resource.Info, o *WaitOptions) (finalObject runtime.Object, done bool, err error)
+
+// RunWait runs the waiting logic
+func (o *WaitOptions) RunWait() error {
+	return o.ResourceFinder.Do().Visit(func(info *resource.Info, err error) error {
+		if err != nil {
+			return err
+		}
+
+		finalObject, success, err := o.ConditionFn(info, o)
+		if success {
+			o.Printer.PrintObj(finalObject, o.Out)
+			return nil
+		}
+		if err == nil {
+			return fmt.Errorf("%v unsatisified for unknown reason", finalObject)
+		}
+		return err
+	})
+}
+
+// IsDeleted is a condition func for waiting for something to be deleted
+func IsDeleted(info *resource.Info, o *WaitOptions) (runtime.Object, bool, error) {
+	endTime := time.Now().Add(o.Timeout)
+	for {
+		gottenObj, err := o.DynamicClient.Resource(info.Mapping.Resource).Namespace(info.Namespace).Get(info.Name, metav1.GetOptions{})
+		if errors.IsNotFound(err) {
+			return info.Object, true, nil
+		}
+		if err != nil {
+			// TODO this could do something slightly fancier if we wish
+			return info.Object, false, err
+		}
+
+		watchOptions := metav1.ListOptions{}
+		watchOptions.FieldSelector = "metadata.name=" + info.Name
+		watchOptions.ResourceVersion = gottenObj.GetResourceVersion()
+		objWatch, err := o.DynamicClient.Resource(info.Mapping.Resource).Namespace(info.Namespace).Watch(watchOptions)
```

Comment:
There is a chance that there will be a delete event between `Get` and `Watch`? i.e. we may miss the delete event and timeout, if the delete event was after we checked that the object is still there, but before we started a watch. We should probably setup watch before running `Get`?

Reply:
What happens if we try to subscribe for events on an object that already doesn't exist? Will there be an error? If so, then the code is fine.

Reply:
Watching from a resource version avoids the risk of missing the delete.  It will send us the delete.

If you watch a thing that doesn't exist, it just waits until it does.  The condition wait does this